File: 015.cpp
class Solution {
public:  
    
    vector<vector<int>> ans; 
    void twosum(vector<int>&nums, int target, int begin){
        int lp = begin, rp = nums.size() -1; 
        if (nums[begin] > target) return ; 
        while (lp < rp) {

            while ( lp < rp && (nums[lp] + nums[rp] < target || (lp != begin &&nums[lp] == nums[lp-1] )) )  lp++;  
            while( lp < rp && (nums[lp] + nums[rp] > target || (rp != nums.size() -1 && nums[rp] == nums[rp+1])))  rp--;  
            if (lp >= rp) return; 
            if (nums[lp] + nums[rp] == target ) {
                ans.push_back({nums[lp], nums[rp], -target}); 
                lp++; 
                rp --; 
            }
            
        }
    }
    vector<vector<int>> threeSum(vector<int>& nums) { 
        if (nums.size() < 3 ) return {}; 
        sort(nums.begin(), nums.end());  
        int i = 1; 
        twosum(nums, -nums[0], i); 
        
        for (i ; i < nums.size() -2; i++){
            if (nums[i] != nums[i-1]){
                twosum(nums, -nums[i], i+1);          
            }
        } 
        
        return ans; 
    } 
};

File: 015.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        self.nums_size = len(nums)      

        if self.nums_size < 3: 
            return []
        
        self.target_sum = -nums[0] 
        self.result = set({})
        self.nums = nums
        
        for i,number in enumerate(nums):
            self.target_sum = -number
            self.findPair(i)
        
        result_size = len(self.result)
        result_array = []
        for i, tup in enumerate(self.result):
            result_array.append(list(tup))
        
        return result_array
            
    def findPair(self,curr_index):
        hash_set = set({})
        for i in range(curr_index+1, self.nums_size): 
            curr_number = self.nums[i]
            target_complement = self.target_sum - curr_number
            
            if target_complement not in hash_set: 
                hash_set.add(curr_number)
            else:
                array_res = sorted([-self.target_sum, curr_number, target_complement])
                self.result.add(tuple(array_res))
